src/components/Select/Select.tsx

```diff
@@ -0,0 +1,357 @@
+import { Listbox } from '@headlessui/react';
+import clsx from 'clsx';
+import type { ReactNode, ElementType } from 'react';
+import React, { useContext } from 'react';
+import styles from './Select.module.css';
+
+import type { ExtractProps } from '../../util/utility-types';
+
+import Icon from '../Icon';
+
+import PopoverContainer from '../PopoverContainer';
+import PopoverListItem from '../PopoverListItem';
+
+export type OptionsAlignType = 'left' | 'right';
+export type VariantType = 'compact' | 'full';
+
+type RenderProp<Arg> = (arg: Arg) => ReactNode;
+type PropsWithRenderProp<RenderPropArg> = {
+  children?: ReactNode | RenderProp<RenderPropArg>;
+  className?: string;
+  as?: ElementType;
+};
+
+type SelectProps = ExtractProps<typeof Listbox> & {
+  /**
+   * Screen-reader text for the select's label.
+   *
+   * When possible, use a visible label through the `labelText` prop or
+   * by passing a <Select.Label> into `chidren`. In rare cases where there's no
+   * visible label, you must provide an `aria-label` for screen readers.
+   * If you pass in an `aria-label`, you don't need `labelText` or <Select.Label>.
+   */
+  'aria-label'?: string;
+  /**
+   * Optional className for additional styling.
+   */
+  className?: string;
+  /**
+   * The style of the select.
+   *
+   * Compact renders select trigger button that is only as wide as the content.
+   */
+  variant?: VariantType;
+  /**
+   * Align select's popover to the left (default) or right of the trigger button.
+   */
+  optionsAlign?: OptionsAlignType;
+  /**
+   * Optional className for additional options menu styling.
+   *
+   * When not using the compact variant, if optionsClassName is provided please
+   * include the width property to define the options menu width.
+   */
+  optionsClassName?: string;
+};
+
+function childrenHaveLabelComponent(children?: ReactNode): boolean {
+  const childrenArray = React.Children.toArray(children);
+  return childrenArray.some((child) => {
+    if (typeof child === 'string' || typeof child === 'number') {
+      return false;
+    } else if (
+      'props' in child &&
+      child.type &&
+      typeof child.type !== 'string' &&
+      child.type?.name === 'SelectLabel'
+    ) {
+      return true;
+    } else if ('props' in child && child.props.children) {
+      return childrenHaveLabelComponent(child.props.children);
+    }
+    return false;
+  });
+}
+
+/**
+ * `import {Select} from "@chanzuckerberg/eds";`
+ *
+ * EDS Select. Used to select one option from a list of options.
+ *
+ * Built on top of the Headless UI Listbox: https://headlessui.dev/react/listbox#basic-example
+ *
+ * TODO: document using class names to alter widths
```

Comment:
Leftover todo?

Reply:
yah sorta? I wasn't sure how this currently works at first, but I did include a test with the result in, so i can address that